sopa/embedding/patches.py

```diff
@@ -0,0 +1,150 @@
+import logging
+from typing import Callable
+
+import dask as da
+import numpy as np
+import tqdm
+from spatial_image import SpatialImage
+from spatialdata import SpatialData, bounding_box_query
+from spatialdata.models import Image2DModel
+from spatialdata.transformations import Scale
+
+from sopa._sdata import get_intrinsic_cs, get_key
+from sopa.segmentation import Patches2D
+
+log = logging.getLogger(__name__)
+
+
+def _get_best_level_for_downsample(
+    level_downsamples: list[float], downsample: float, epsilon: float = 0.01
+) -> int:
+    """Return the best level for a given downsampling factor"""
+    if downsample <= 1.0:
+        return 0
+    for level, ds in enumerate(level_downsamples):
+        if ds > downsample + epsilon:
+            return level - 1
+    return len(level_downsamples) - 1
+
+
+def _get_extraction_parameters(
+    tiff_metadata: dict, target_magnification: int, patch_width: int
+) -> tuple[int, int, int, bool]:
+    """
+    Given the metadata for the slide, a target magnification and a patch width,
+    it returns the best scale to get it from (level), a resize factor (resize_factor)
+    and the corresponding patch size at scale0 (patch_width)
+    """
+    if tiff_metadata["properties"].get("tiffslide.objective-power"):
+        downsample = (
+            int(tiff_metadata["properties"].get("tiffslide.objective-power")) / target_magnification
+        )
+    elif tiff_metadata["properties"].get("tiffslide.mpp-x"):
+        obj2mpp = {80: 0.125, 40: 0.25, 20: 0.5, 10: 1.0, 5: 2.0}
+        mppdiff = []
```

Comment:
What is the purpose of `mppdiff`? It seems to be also unused

Reply:
Hmm that is a good question. Something looks weird here. I will take a closer look.

Reply:
Ok thanks!